app/models/user.rb

```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  has_secure_password
```

Comment:
tìm hiểu về has_secure_password chưa kĩ lắm, đọc lại nhé

Reply:
vâng

Reply:
has_secure_password cung cấp:
+ Phương thức để thiết lập và xác thực (authenticate) đối với password BCrypt
yêu cầu phải có thuộc tính password_digest
+ Validations sẽ được thêm tự động
+ Password phải tồn tại
+ Chiều dài Password phải <= 72 kí tự
+ Confirmation password (Sử dụng thuộc tính password_confirmation)

Reply:
Validate những gì và validate khi nào?

Reply:
Đảm bảo User phải có Password
Password ít nhất là 6 kí tự
password_digest không được để trống
Validate khi password phải khớp với password_confirmation

Reply:
password phải khớp với confirmation là 1 cái validate rồi.
còn cái validate ở thời điểm nào thì e chưa đọc được từ khóa đó.

Reply:
các method dưới đây sẽ gọi Validate
+ create
+ create!
+ save
+ save!
+ update
+ update!

các method dưới đây sẽ không gọi Validate
+ decrement!
+ decrement_counter
+ increment!
+ increment_counter
+ toggle!
+ touch
+ update_all
+ update_attribute
+ update_column
+ update_columns
+ update_counters

Reply:
ý a là các validation mà has_secure_password cung cấp cơ

Reply:
Không được bỏ trống -> presence: true

Các giá trị boolean -> inclusion: {in: [true, false]}

Checkboxs -> acceptance: true

Phải để trống -> absence: true

Tính duy nhất -> uniqueness: true

Tính đồng nhất -> confirmation: true

inclusion -> inclusion: {in: %w(draft publish private)}
(Kiểm tra giá trị có nằm trong tập hợp cho trước không.)

exclusion -> exclusion: {in: %w(VN JP ID US)}
(Kiểm tra giá trị không được nằm trong tập hợp cho trước.)

Độ dài -> 
length: { minimum: 1 }       # 1 kí tự trở lên
length: { maximum: 10 }      # 10 kí tự trở xuống
length: { in: 1..10 }        # 1 kí tự trở lên, 10 kí tự trở xuống
length: { is: 8 }         # chỉ 8

Validate kiểu số -> numericality: true

Email -> VALID_EMAIL_REGEX = /A[w+-.]+@[a-zd-.]+.[a-z]+z/i
presence: true, uniqueness: true, format: { with: VALID_EMAIL_REGEX }

Reply:
Validate được thực thi ở thời điểm User đã được tạo
@user = User.new(name: "", email: "") hoặc User.create(...)